components/script/dom/bindings/str.rs

```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
     })
 }
 
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let lang_ranges: Vec<&str> = range.split(',').collect();
+
+    lang_ranges.iter().any(|&lang_range| {
+        // step 1
+        let range_subtags: Vec<&str> = lang_range.split('\x2d').collect();
+        let tag_subtags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(lang_range) || lang_range.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        let mut range_iter = range_subtags.iter();
+        let mut tag_iter = tag_subtags.iter();
+
+        // step 2
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
+        if let (Some(range_subtag), Some(tag_subtag)) = (range_iter.next(), tag_iter.next()) {
+            if !(range_subtag.eq_ignore_ascii_case(tag_subtag) || range_subtag.eq_ignore_ascii_case("*")) {
+                return false;
+            }
+        }
+
+        // step 3
+        for range_subtag in range_iter {
+            // step 3a
+            if range_subtag.eq_ignore_ascii_case("*") {
+                continue;
+            }
+            loop {
```

Comment:
this is interesting. 
This pretty much looks good. But the problem is when you have the tag length as `2` ( actually when range and tag is equal in length ).

consider this, 
tag -> `en-GB`
range -> `en-US`

if we don't have that inner loop. 
step 2 - tag and range will match. (`en`)

step 3 -  (first loop)
    a - not applicable
    b - not applicable
    c - will fail
    d - will fail
    e - will move to the next subtag. which will be `None`

Step 3 will finish here since there will not be any other elements in range

Step 4 - returns true ( which is wrong )

In other scenarios it does make sense. am I missing anything here.